magellan-library/src/main/java/com/wealthfront/magellan/navigation/LinearNavigator.kt

```diff
@@ -0,0 +1,187 @@
+package com.wealthfront.magellan.navigation
+
+import android.content.Context
+import android.view.View
+import androidx.annotation.VisibleForTesting
+import com.wealthfront.magellan.Direction
+import com.wealthfront.magellan.Direction.BACKWARD
+import com.wealthfront.magellan.Direction.FORWARD
+import com.wealthfront.magellan.NavigationType
+import com.wealthfront.magellan.NavigationType.GO
+import com.wealthfront.magellan.NavigationType.SHOW
+import com.wealthfront.magellan.ScreenContainer
+import com.wealthfront.magellan.core.Navigable
+import com.wealthfront.magellan.lifecycle.LifecycleAwareComponent
+import com.wealthfront.magellan.lifecycle.LifecycleState.Created
+import com.wealthfront.magellan.lifecycle.LifecycleState.Destroyed
+import com.wealthfront.magellan.lifecycle.LifecycleState.Resumed
+import com.wealthfront.magellan.lifecycle.LifecycleState.Shown
+import com.wealthfront.magellan.lifecycle.LifecycleStateMachine
+import com.wealthfront.magellan.transitions.DefaultTransition
+import com.wealthfront.magellan.view.whenMeasured
+import java.util.Deque
+import java.util.LinkedList
+
+class LinearNavigator(
+  private val container: () -> ScreenContainer
+) : Navigator, LifecycleAwareComponent() {
+
+  private val lifecycleStateMachine = LifecycleStateMachine()
+
+  @VisibleForTesting
+  override val backStack: LinkedList<NavigationEvent> = LinkedList()
+
+  private var ghostView: View? = null
+  private var containerView: ScreenContainer? = null
+
+  private val currentNavigable: Navigable?
+    get() = backStack.peek()?.navigable
+
+  private val context: Context?
+    get() = currentState.context
+
+  override fun onCreate(context: Context) {
+    currentNavigable?.let {
+      lifecycleStateMachine.transitionBetweenLifecycleStates(it, Destroyed, Created(context))
+    }
+    lifecycleStateMachine.transitionBetweenLifecycleStates(
+      backStack.navigables(),
+      Destroyed,
+      Created(context))
+  }
+
+  override fun onShow(context: Context) {
+    containerView = container()
+    currentNavigable?.let {
+      containerView!!.addView(it.view!!)
+    }
+  }
+
+  override fun onDestroy(context: Context) {
+    currentNavigable?.let {
+      lifecycleStateMachine.transitionBetweenLifecycleStates(it, Created(context), Destroyed)
+    }
+    lifecycleStateMachine.transitionBetweenLifecycleStates(
+      backStack.navigables(),
+      Created(context),
+      Destroyed)
```

Comment:
When we introduce `navigator.showAbove` or whatever we'll call it, we should also introduce a way to continue to be attached to the lifecycle owner but not receive events past a certain point. For example, the screens in our backstack could still be attached, just not receive any events past `create`, or a screen behind one that's "shown above" it would still be attached, just not receive any events past `start`/`show`

Reply:
Yup. There's a ticket for that.